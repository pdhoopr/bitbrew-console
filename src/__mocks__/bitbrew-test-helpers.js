import diff from 'jest-diff';
import ShallowRenderer from 'react-test-renderer/shallow';

export function diffComponents(baseComponent, compareComponent) {
  const changes = diff(baseComponent, compareComponent, {
    aAnnotation: `<BaseComponent />`,
    bAnnotation: `<CompareComponent />`,
    contextLines: 3,
    expand: false,
  });
  return `Diff:\n${changes}`;
}

export function shallow(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

export const fixtures = {
  accessToken:
    'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE1MjUxMTUzNTAsImV4cCI6MTU1NjY1MTM1MCwiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSJ9.4O-W8aeOOUFU8fkGTAFaAu9D9nK_l61yPomNG1Qkny8',
  orgs: {
    danlaw: {
      id: 'bd006616-3bd2-4783-a84d-b0fa4ed81135',
      properName: 'Danlaw',
    },
    azuga: {
      id: '8d9ec637-f3bb-4c46-83ee-11764d09f0b2',
      properName: 'Azuga',
    },
    bitbrew: {
      id: '94d9b0ca-a758-4502-8dcd-41d883e38714',
      properName: 'BitBrew',
    },
  },
  users: {
    bill: {
      email: 'bill@bitbrew.com',
      name: 'Bill Monaghan',
    },
    cynthia: {
      email: 'cynthia@bitbrew.com',
      name: 'Cynthia Hamilton',
    },
    alex: {
      email: 'bill@bitbrew.com',
      name: 'Alex Li',
    },
  },
};
