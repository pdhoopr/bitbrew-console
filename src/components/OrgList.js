import { PropTypes as MobxPropTypes, observer } from 'mobx-react';
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import SearchIcon from '../graphics/search-icon.svg';
import Search from '../models/Search';
import { RaisedInput } from './Inputs';

const SearchBar = styled.label`
  display: block;
  margin-top: var(--size-32);
  position: relative;
`;

const Icon = styled(SearchIcon)`
  fill: var(--color-dark-grey);
  height: var(--size-24);
  margin-left: var(--size-16);
  margin-top: 6px;
  position: relative;
  width: var(--size-24);
  z-index: 1;
`;

const Input = RaisedInput.extend`
  left: 0;
  padding-left: calc(var(--size-16) + var(--size-24) + var(--size-16));
  position: absolute;
  top: 0;
`;

const Section = styled.section`
  margin-top: var(--size-32);
`;

const Title = styled.h2`
  font-size: var(--size-24);
  font-weight: var(--weight-normal);
  line-height: var(--size-32);
`;

const searchByOrgName = orgs => query =>
  orgs.filter(org => new RegExp(query, 'i').test(org.properName));

class OrgList extends React.Component {
  search = Search.create({}, { onSearch: searchByOrgName(this.props.orgs) });

  render() {
    return (
      <>
        <SearchBar htmlFor="search">
          <Icon />
          <Input
            id="search"
            value={this.search.query}
            onChange={this.search.update}
            type="search"
            placeholder="Search by organization name"
          />
        </SearchBar>
        {this.search.results.map(org => (
          <Section key={org.id}>
            <Title data-testid="org-name">{org.properName}</Title>
          </Section>
        ))}
      </>
    );
  }
}

OrgList.propTypes = {
  orgs: MobxPropTypes.arrayOrObservableArrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      properName: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default observer(OrgList);
