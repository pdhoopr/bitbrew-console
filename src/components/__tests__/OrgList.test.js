import { diffComponents, fixtures, shallow } from 'bitbrew-test-helpers';
import React from 'react';
import { Simulate, render } from 'react-testing-library';
import OrgList from '../OrgList';

test('renders DOM for the org list without any orgs', () => {
  const dom = shallow(<OrgList orgs={[]} />);
  expect(dom).toMatchSnapshot();
});

test('renders different DOM for the org list with orgs', () => {
  const base = shallow(<OrgList orgs={[]} />);
  const compare = shallow(<OrgList orgs={Object.values(fixtures.orgs)} />);
  const diff = diffComponents(base, compare);
  expect(diff).toMatchSnapshot();
});

test('can filter the org list with a case-insensitive search query', async () => {
  const orgs = Object.values(fixtures.orgs);
  const query = 'A';
  const results = [fixtures.orgs.azuga, fixtures.orgs.danlaw];
  const { getAllByTestId, getByPlaceholderText } = render(
    <OrgList orgs={orgs} />,
  );
  expect(getAllByTestId('org-name').length).toBe(orgs.length);
  const search = getByPlaceholderText('Search by organization name');
  search.value = query;
  Simulate.change(search);
  const allOrgNames = getAllByTestId('org-name').map(
    orgName => orgName.textContent,
  );
  expect(allOrgNames.length).toBe(results.length);
  results.forEach(result => {
    expect(allOrgNames).toContain(result.properName);
  });
});
