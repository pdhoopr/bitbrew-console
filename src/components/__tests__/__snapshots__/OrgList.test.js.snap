// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DOM for the org list without any orgs 1`] = `
<UNDEFINED>
  <styled.label
    htmlFor="search"
  >
    <styled.svg />
    <styled.input
      id="search"
      onChange={[Function]}
      placeholder="Search by organization name"
      type="search"
      value=""
    />
  </styled.label>
</UNDEFINED>
`;

exports[`renders different DOM for the org list with orgs 1`] = `
Diff:
- <BaseComponent />
+ <CompareComponent />

@@ -11,4 +11,25 @@
        value=""
      />
    </styled.label>
+   <styled.section>
+     <styled.h2
+       data-testid="org-name"
+     >
+       Danlaw
+     </styled.h2>
+   </styled.section>
+   <styled.section>
+     <styled.h2
+       data-testid="org-name"
+     >
+       Azuga
+     </styled.h2>
+   </styled.section>
+   <styled.section>
+     <styled.h2
+       data-testid="org-name"
+     >
+       BitBrew
+     </styled.h2>
+   </styled.section>
  </UNDEFINED>
`;
